04-cs-fundamentals/classroom/Finite-State-Machines.md

```diff
@@ -0,0 +1,182 @@
+# Finite State Machines
+
+People have been struggling with the theory of computation long before computers existed.  Mathematics is a powerful tool and mathematicians have long sought to answer, "What can I compute?"  There are problems that **cannot** be solved computationally and there are problems we cannot solve in a reasonable (non-exponential) amount of time.
+
+The study of this field is called [**Automata Theory**](https://en.wikipedia.org/wiki/Automata_theory).  It's really about what can and cannot be computed and how to represent it.  
+
+We will only touch on the 1st level of Automata theory, **Finite State Machines** (FSMs).  This is one of several ways to model computation, and a good place to start.
+
+## Learning Goals
+
+By the end of this lesson you should be able to:
+
+- Explain what a Finite State Machine (FSM) is
+- Identify the languages a FSM can solve
+- Design a FSM to solve for a specific language pattern
+- Explain what kinds of problems a DFA cannot solve
+
+## What is a Finite State Machine
+
+A Finite State Machine is a relatively simple model of computation.  An FSA is a model of a machine which can be in exactly *one* of a limited (i.e. finite) set of states.  The machine can take in input which causes the machine to change from one state to another.  Moving from one state to another is called a **transition**.  An FSA always has one state in which it starts and one or more states which are valid end-points.  If a FSA takes in input and ends at a valid end-point, the input is considered accepted.  Otherwise the input is considered invalid and rejected.  
```

Comment:
You use FSA and FSM throughout the lesson.  I think using FSM everywhere is probably clearer, though it would require renaming some images and if you think you're likely to say FSA or DFA out loud it may be worth defining those terms. 

```suggestion
A Finite State Machine (FSM) is a relatively simple model of computation.  An FSM is a model of a machine which can be in exactly *one* of a limited (i.e. finite) set of states.  The machine can take in input which causes the machine to change from one state to another.  Moving from one state to another is called a **transition**.  An FSM always has one state in which it starts and one or more states which are valid end-points.  If an FSM takes in input and ends at a valid end-point, the input is considered accepted.  Otherwise the input is considered invalid and rejected.  
```